ipa-core/src/protocol/ipa_prf/shuffle/malicious.rs

```diff
@@ -27,16 +29,110 @@ use crate::{
+///
+/// ## Errors
+/// Propagates network, multiplication and conversion errors from sub functions.
+///
+/// ## Panics
+/// Panics when `S::Bits + 32 != B::Bits` or type conversions fail.
+pub async fn malicious_shuffle<C, S, B, I>(
+    ctx: C,
+    shares: I,
+) -> Result<Vec<AdditiveShare<S>>, Error>
+where
+    C: Context,
+    S: BooleanArray,
+    B: BooleanArray,
+    I: IntoIterator<Item = AdditiveShare<S>>,
+    I::IntoIter: ExactSizeIterator,
+    <I as IntoIterator>::IntoIter: Send,
+    for<'a> &'a B: Add<B, Output = B>,
+    for<'a> &'a B: Add<&'a B, Output = B>,
+    Standard: Distribution<B>,
+{
+    // compute amount of MAC keys
+    let amount_of_keys: usize = (usize::try_from(S::BITS).unwrap() + 31) / 32;
+    // // generate MAC keys
+    let keys = (0..amount_of_keys)
+        .map(|i| ctx.prss().generate_fields(RecordId::from(i)))
+        .map(|(left, right)| AdditiveShare::new(left, right))
+        .collect::<Vec<AdditiveShare<Gf32Bit>>>();
+
+    // compute and append tags to rows
+    let shares_and_tags: Vec<AdditiveShare<B>> =
+        compute_and_add_tags(ctx.narrow(&OPRFShuffleStep::GenerateTags), &keys, shares).await?;
+
+    // shuffle
+    let (shuffled_shares, messages) = shuffle(
+        ctx.narrow(&OPRFShuffleStep::ShuffleProtocol),
+        shares_and_tags,
+    )
+    .await?;
+
+    // verify the shuffle
+    verify_shuffle::<_, S, B>(
+        ctx.narrow(&OPRFShuffleStep::VerifyShuffle),
+        &keys,
+        &shuffled_shares,
+        messages,
+    )
+    .await?;
+
+    // truncate tags from output_shares
+    Ok(truncate_tags(&shuffled_shares))
+}
+
+/// This function truncates the tags from the output shares of the shuffle protocol
+///
+/// ## Panics
+/// Panics when `S::Bits > B::Bits`.
+fn truncate_tags<S, B>(shares_and_tags: &[AdditiveShare<B>]) -> Vec<AdditiveShare<S>>
+where
+    S: BooleanArray,
+    B: BooleanArray,
+{
+    let tag_offset = usize::try_from((S::BITS + 7) / 8).unwrap();
+    shares_and_tags
+        .iter()
+        .map(|row_with_tag| {
+            AdditiveShare::new(
+                split_row_and_tag(row_with_tag.left(), tag_offset).0,
+                split_row_and_tag(row_with_tag.right(), tag_offset).0,
+            )
+        })
+        .collect()
+}
+
+/// This function splits a row with tag into
+/// a row without tag and a tag
+///
+/// ## Panics
+/// Panics when the lengths are incorrect:
+/// `S` in bytes needs to be equal to `tag_offset`.
+/// `B` in bytes needs to be equal to `tag_offset + 4`.
+fn split_row_and_tag<S: BooleanArray, B: BooleanArray>(
+    row_with_tag: B,
+    tag_offset: usize,
+) -> (S, Gf32Bit) {
+    let mut buf = GenericArray::default();
+    row_with_tag.serialize(&mut buf);
+    (
+        S::deserialize(GenericArray::from_slice(&buf.as_slice()[0..tag_offset])).unwrap(),
+        Gf32Bit::deserialize(GenericArray::from_slice(&buf.as_slice()[tag_offset..])).unwrap(),
```

Comment:
I'm not sure the unwrap of `S::deserialize` is okay. It will fail if `S` is not a multiple of bytes and the padding bits are non-zero. (Which means a server panic reachable from external input.)

Possibly it can be changed to `unwrap_or_default()`, and in the case that a row is replaced with the default (zero), the verification fails whp. If that's sound, please note the justification in a comment.